design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+example, given the Smithy model:
+
+```smithy
+@enum([
+    { name: "Spanish", value: "es" },
+    { name: "English", value: "en" },
+    { name: "Japanese", value: "jp" },
+])
+string Language
+```
+
+the code the client generates when deserializing a string from a JSON document
+into the `Language` enum is (excerpt):
+
+```rust
+...
+match key.to_unescaped()?.as_ref() {
+    "language" => {
+        builder = builder.set_language(
+            aws_smithy_json::deserialize::token::expect_string_or_null(
+                tokens.next(),
+            )?
+            .map(|s| {
+                s.to_unescaped()
+                    .map(|u| crate::model::Language::from(u.as_ref()))
+            })
+            .transpose()?,
+        );
+    }
+    _ => aws_smithy_json::deserialize::token::skip_value(tokens)?,
+}
+...
+```
+
+Note how the `String` gets converted to the enum via `Language::from()`.
+
+```rust
+impl std::convert::From<&str> for Language {
+    fn from(s: &str) -> Self {
+        match s {
+            "es" => Language::Spanish,
+            "en" => Language::English,
+            "jp" => Language::Japanese,
+            other => Language::Unknown(other.to_owned()),
+        }
+    }
+}
+```
+
+For constrained shapes we would do the same to parse the inner deserialized
+value into the wrapper tuple struct, except for these differences:
+
+1. For enums, the client generates an `Unknown` variant that "contains new
+   variants that have been added since this code was generated". The server
+   does not need such a variant ([#1187]).
+2. Conversions into the tuple struct are fallible (`try_from()` instead of
+   `from()`). These errors will result in a `MyStructValidationError`, wrapped
+   in the runtime type error
+   `aws_smithy_http_server::rejection::ConstraintViolation` rejection. The
+   deseralizer function will wrap this rejection in an
+   [`aws_smithy_http_server::rejection::Deserialize`] rejection, to indicate
+   that deserialization failed as a result of a constraint violation.
+   Eventually this rejection will be wrapped in an
+   [`aws_smithy_http_server::rejection::SmithyRejection`], our parent rejection
+   enum type for all errors that can occur in the smithy-rs service framework.
+   When serializing this error into an HTTP response, the error message that
+   naturally arises by calling `std::fmt::Display::fmt()` will trace what
+   exactly went wrong for the caller client.
+
+`length` trait
+--------------
+
+We will enforce the length constraint by calling `len()` on Rust's `Vec`
+(`list` and `set` shapes), `HashMap` (`map` shapes), `String` (`string`
+shapes), and our [`aws_smithy_types::Blob`] (`bytes` shapes).
+
+`pattern` trait
+---------------
+
+The [`pattern` trait]
+
+> restricts string shape values to a specified regular expression.
+
+We will implement this by using the `regex`'s crate [`is_match`]. We will use
+[`lazy_static!`] to compile the regex only the first time it is required.
```

Comment:
[`once_cell`](https://lib.rs/crates/once_cell) is now preferred to `lazy_static`